### ISSUE TITLE
[Feature]: Implement RSLAD-style robust distillation in Compress phase

### ISSUE THREAD

**Issue by Adit-Jain-srm** [Adit-Jain-srm/NightmareNet#110]
## Problem

Paper Section 3.5 defines the compression loss as:

```
L_compress = alpha * L_task + (1-alpha) * KL(p_teacher || p_student) + beta * KL(p_teacher_adv || p_student_adv)
```

The code (`nightmarenet/training/phases.py:429`, `CompressionPhase.run()`) only implements magnitude pruning + standard fine-tuning. There is no teacher model, no KL divergence, and no adversarial examples during compression. This is the most critical paper-code mismatch blocking publication.

Builds on #47. Related: Zi et al. 2021 "Revisiting Adversarial Robustness Distillation from the Perspective of Robust Optimization" (RSLAD).

**Relevant files:**
- `nightmarenet/training/phases.py` lines 429-530 (CompressionPhase)
- `nightmarenet/compression/pruning.py` (MagnitudePruner - currently the only compression method)
- `docs/research/paper-draft.md` Section 3.5 (describes the intended formula)

## Proposed Solution

- Before pruning, snapshot the current model as "teacher"
- After pruning, run a distillation loop where the pruned "student" minimizes:
  - Task loss on clean data
  - KL divergence from teacher on clean data
  - KL divergence from teacher on adversarially-perturbed data (RSLAD key insight)
- Config: `compression.method: distillation|pruning`, `compression.distillation_alpha: 0.7`, `compression.distillation_temperature: 4.0`
- The student model should be smaller AND retain adversarial decision boundaries

## Acceptance Criteria

- [ ] Distillation loss matches the paper formula (task + clean KL + adversarial KL)
- [ ] Teacher model is a frozen copy of pre-compression model
- [ ] Student model has fewer parameters than teacher (via pruning before distillation)
- [ ] Adversarial examples generated during distillation (not just clean data)
- [ ] Config supports both `pruning` (existing) and `distillation` (new) modes
- [ ] Unit tests cover the distillation training loop
- [ ] Robustness of distilled student measured and compared to pruned-only baseline

## Scope

Large - new training paradigm in compression phase. Requires understanding of knowledge distillation + adversarial training literature. Multi-day.

---

> **Important:** Please read [CONTRIBUTING.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/CONTRIBUTING.md) in full before requesting assignment. Assignment requests without a technical approach will not be considered. PRs demonstrating thorough research, benchmarks, or creative approaches beyond minimum criteria may earn bonus XP labels.

**Comment by Diwakar-odds** [comment]
Hi @Adit-Jain-srm, I would love to tackle this critical paper-blocker issue under ECSoC '26. The knowledge distillation paradigm proposed in Section 3.5 is extremely interesting, and I have prepared a detailed approach to integrate RSLAD-style distillation into the Compression Phase.

### Problem Description
Currently, the `CompressionPhase` only relies on simple magnitude pruning and standard fine-tuning. To properly match the paper's formula (Section 3.5), we need to implement a robust knowledge distillation loop where the student minimizes the task loss, the KL divergence against the teacher on clean data, and crucially, the KL divergence on adversarially-perturbed data.

### Implementation Plan
1. **Teacher Model State**: Before any pruning occurs in `CompressionPhase.run()`, I will clone and freeze the current model (`copy.deepcopy` and `requires_grad=False`). This serves as the teacher model.
2. **Student Model Initialization**: Execute the existing `MagnitudePruner` logic to compress the network. This pruned model becomes our student.
3. **Robust Distillation Loop**: 
   - I will implement a new training loop specifically for the distillation mode inside `nightmarenet/training/phases.py`.
   - For each batch, calculate the clean output for both teacher and student to compute the clean KL divergence.
   - Generate adversarial perturbations (e.g., via PGD/FGSM logic borrowed from the Nightmare phase) for the batch.
   - Compute the adversarial KL divergence between the teacher and student on these perturbed inputs.
   - Sum the components: `L_task + (1-alpha)*KL_clean + beta*KL_adv`.
4. **Configuration Layer**: Update the configuration schema (or config dataclasses) to support `compression.method`, `distillation_alpha`, `distillation_beta` (for the adversarial term multiplier), and `distillation_temperature`.
5. **Testing**: I will add unit tests in `tests/test_compression.py` that mock the adversarial generation and assert the multi-component loss calculation matches expectations.

### Files to Modify
- `nightmarenet/training/phases.py` (Main logic for `CompressionPhase`)
- `nightmarenet/config/schema.py` or wherever configuration logic resides.
- `tests/test_compression.py` (New tests)

### Questions for the Maintainer
- What adversarial attack logic should we use for generating `p_student_adv`? Should we reuse the existing `NightmarePhase` distortion registry, or write a lightweight, dedicated PGD attacker inside the distillation loop to save on compute overhead?

### Expected Outcome
The `CompressionPhase` will correctly mirror the Section 3.5 math, allowing the framework to produce compressed student models that retain the adversarial robustness of their teachers. This resolves the paper-code mismatch!

Please let me know if I can take this assignment!

**Comment by BikramMondal5** [comment]
Hi! I'd like to work on this issue if it's still available.

### Approach
- Update `CompressionPhase` to support a new `distillation` mode while preserving the existing `pruning` workflow.
- Snapshot the pre-pruning model as a frozen teacher, prune the model to obtain the student, and implement the paper's compression objective combining task loss, clean KL divergence, and adversarial KL divergence.
- Generate adversarial examples during the distillation phase to align the student's robust decision boundaries with the teacher, following the RSLAD approach.
- Add configurable parameters for `compression.method`, `compression.distillation_alpha`, and `compression.distillation_temperature`.
- Add unit tests covering the distillation training loop and validate that the student model is pruned, the teacher remains frozen, and robustness is measured against the pruning-only baseline.

Looking forward to contributing!

**Comment by Adit-Jain-srm** [comment]
Assigned @Diwakar-odds

**Comment by Adit-Jain-srm** [comment]
> What adversarial attack logic should we use for generating p_student_adv? Should we reuse the existing NightmarePhase distortion registry, or write a lightweight, dedicated PGD attacker inside the distillation loop to save on compute overhead?

No need to save on compute overhead @Diwakar-odds
